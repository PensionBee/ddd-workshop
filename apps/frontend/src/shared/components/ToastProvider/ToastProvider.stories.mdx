import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";
import ToastProvider from "./ToastProvider";

<Meta
  title="Shared/Toast Notifications/Docs"
  component={ToastProvider}
  parameters={{
    viewMode: "docs",
    previewTabs: {
      canvas: { hidden: true },
    },
  }}
/>

# Toast Notifications

## Usage & Customisation

### Success

#### Usage

```js
import { emitToastSuccess } from "@shared/utils/toast";

emitToastSuccess("Event was successful!");
```

#### Example

<Canvas>
  <Story id="shared-toast-notifications--success" />
</Canvas>

### Error

#### Usage

```js
import { emitToastError } from "@shared/utils/toast";

emitToastError("Event had an error!")
```

#### Example

<Canvas>
  <Story id="shared-toast-notifications--error" />
</Canvas>

### Info

#### Usage

```js
import { emitToastInfo } from "@shared/utils/toast";

emitToastInfo("Some useful information.")
```

#### Example

<Canvas>
  <Story id="shared-toast-notifications--info" />
</Canvas>

### Promise

#### Usage

```js
import { emitToastPromise } from "@shared/utils/toast";

emitToastPromise(asyncWaitThenSuccess(), {
  pending: "Loading... (5 Seconds)",
  success: "Success!",
  error: "Failed!",
})
```

#### Example Success

<Canvas>
  <Story id="shared-toast-notifications--promise-success" />
</Canvas>

#### Example Error

<Canvas>
  <Story id="shared-toast-notifications--promise-error" />
</Canvas>
