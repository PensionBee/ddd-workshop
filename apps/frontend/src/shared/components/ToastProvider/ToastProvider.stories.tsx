import { Button, type ButtonProps } from "@shared/components/Button";
import { Box } from "@shared/components/Box";
import {
  emitToastError,
  emitToastInfo,
  emitToastPromise,
  emitToastSuccess,
} from "@shared/utils/toast";
import type { ComponentMeta, ComponentStory } from "@storybook/react";
import { ToastProvider } from "./ToastProvider";

export default {
  title: "Shared/Toast Notifications",
  component: Button,
  decorators: [
    (Story) => (
      <>
        <Box d="flex" p={20}>
          <Story />
          <ToastProvider />
        </Box>
      </>
    ),
  ],
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof Button>;

const NotificationTemplate: ComponentStory<typeof Button> = (
  args: ButtonProps
) => {
  return <Button {...args} />;
};

export const Success = NotificationTemplate.bind({});
Success.args = {
  onClick: () => emitToastSuccess("Event was successful!"),
  children: "Emit Success Toast",
};
export const Error = NotificationTemplate.bind({});
Error.args = {
  onClick: () => emitToastError("Event had an error!"),
  children: "Emit Error Toast",
};
export const Info = NotificationTemplate.bind({});
Info.args = {
  onClick: () => emitToastInfo("Some useful information."),
  children: "Emit Info Toast",
};

const asyncWaitThenSuccess = () =>
  new Promise((resolve) => {
    setTimeout(() => resolve("Success"), 5000);
  });

const handlePromiseSuccessButtonClick = async () => {
  return emitToastPromise(asyncWaitThenSuccess(), {
    pending: "Loading... (5 Seconds)",
    success: "Success!",
    error: "Failed!",
  });
};

export const PromiseSuccess = NotificationTemplate.bind({});
PromiseSuccess.args = {
  onClick: handlePromiseSuccessButtonClick,
  children: "Emit Promise Success",
};

const asyncWaitThenFail = () =>
  new Promise((_, reject) => {
    setTimeout(() => reject("Failed"), 5000);
  });

const handlePromiseErrorButtonClick = async () => {
  return emitToastPromise(asyncWaitThenFail(), {
    pending: "Loading... (5 Seconds)",
    success: "Success!",
    error: "Failed!",
  });
};

export const PromiseError = NotificationTemplate.bind({});
PromiseError.args = {
  onClick: handlePromiseErrorButtonClick,
  children: "Emit Promise Fail",
};
