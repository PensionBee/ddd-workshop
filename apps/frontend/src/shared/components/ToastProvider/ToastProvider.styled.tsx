import { theme } from "@shared/styling/theme";
import { bodySmallM } from "@shared/styling/typography";
import { ToastContainer } from "react-toastify";
import styled from "styled-components";
import "react-toastify/dist/ReactToastify.css";

export const StyledToastContainer = styled(ToastContainer)`
  --toastify-color-light: ${theme.colors.white};
  --toastify-color-dark: ${theme.colors.black};
  --toastify-color-info: ${theme.colors.teal100};
  --toastify-color-success: ${theme.colors.green100};
  --toastify-color-warning: ${theme.colors.orange100};
  --toastify-color-error: ${theme.colors.red100};
  --toastify-color-transparent: rgba(255, 255, 255, 0.7);

  --toastify-icon-color-info: var(--toastify-color-info);
  --toastify-icon-color-success: var(--toastify-color-success);
  --toastify-icon-color-warning: var(--toastify-color-warning);
  --toastify-icon-color-error: var(--toastify-color-error);

  --toastify-toast-width: 320px;
  --toastify-toast-background: ${theme.colors.white};
  --toastify-toast-min-height: 64px;
  --toastify-toast-max-height: 800px;
  --toastify-font-family: sans-serif;
  --toastify-z-index: 9999;

  --toastify-text-color-light: ${theme.colors.black};
  --toastify-text-color-dark: ${theme.colors.white};

  //Used only for colored theme
  --toastify-text-color-info: ${theme.colors.white};
  --toastify-text-color-success: ${theme.colors.white};
  --toastify-text-color-warning: ${theme.colors.white};
  --toastify-text-color-error: ${theme.colors.white};

  --toastify-spinner-color: ${theme.colors.black};
  --toastify-spinner-color-empty-area: ${theme.colors.white};

  // Used when no type is provided
  // toast("**hello**")
  --toastify-color-progress-light: ${theme.colors.yellow};
  // Used when no type is provided
  --toastify-color-progress-dark: ${theme.colors.yellow};
  --toastify-color-progress-info: var(--toastify-color-info);
  --toastify-color-progress-success: var(--toastify-color-success);
  --toastify-color-progress-warning: var(--toastify-color-warning);
  --toastify-color-progress-error: var(--toastify-color-error);

  .Toastify {
    &__toast-body {
      ${bodySmallM}
    }
  }
`;
