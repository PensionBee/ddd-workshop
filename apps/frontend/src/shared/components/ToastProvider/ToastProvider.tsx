import { StyledToastContainer } from "./ToastProvider.styled";

export const ToastProvider: React.FC = () => {
  return (
    <StyledToastContainer
      position="bottom-left"
      autoClose={5000}
      hideProgressBar={false}
      newestOnTop={false}
      closeOnClick
      rtl={false}
      pauseOnFocusLoss
      draggable
      pauseOnHover
    />
  );
};
