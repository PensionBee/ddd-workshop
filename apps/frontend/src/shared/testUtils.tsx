import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
