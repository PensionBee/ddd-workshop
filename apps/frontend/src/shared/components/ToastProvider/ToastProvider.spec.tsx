import { render, vi } from "@shared/testUtils";
import { ToastProvider } from "./ToastProvider";

describe("ToastProvider", () => {
  describe("Default", () => {
    it("should render correctly", () => {
      const { asFragment } = render(<ToastProvider />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
