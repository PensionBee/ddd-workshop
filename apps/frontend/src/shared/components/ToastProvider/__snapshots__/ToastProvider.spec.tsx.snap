// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastProvider Default should render correctly 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
</DocumentFragment>
`;
